### PR TITLE
Task form for 'add new task'

### PR DESCRIPTION
**Pull request by PeterAhern**
## Naming Rules

Name your PR like this: ISSUENUMBER-TITLE-YOURNAME

## Description

I've added a task form, that the user can input task details to be saved

## Related to

Make sure you include the issue number with a hash sign # so Github can link this PR to the right issue, like this:

Fixes #1

## Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have carefully reviewed my own code
- [ ] I have commented my code
- [ ] I have updated any documentation

### CODE REVIEW

File: client/src/App.js [hunk 1]
```diff
@@ -1,13 +1,16 @@
 import { Route, Routes } from "react-router-dom";
 import MyPlan from "./pages/MyPlan";
 import Landing from "./pages/Landing";
-import Card from "./Components/UI/Card/Card";
+// import Card from "./Components/UI/Card/Card";
+
+// import DateSelector from "./Components/TaskForm/DateSelector";
```
Comment (textbook) [inline client/src/App.js:6]: I would recommend not leaving commented-out code (or whole files, like `client/src/Components/TaskForm/Modal.js`) in your pull requests

File: package.json [hunk 2]
```diff
@@ -69,10 +69,15 @@
...
 		"@supabase/supabase-js": "^1.31.1",
+		"bootstrap": "^5.1.3",
 		"express": "^4.17.2",
 		"helmet": "^5.0.2",
+		"jquery": "^3.6.0",
 		"morgan": "^1.10.0",
 		"pg": "^8.7.3",
+		"popper.js": "^1.16.1",
+		"react-bootstrap": "^2.2.0",
+		"react-datepicker": "^4.7.0",
```
Comment (textbook) [inline package.json:80]: These packages only seem to be used by the React client build, therefore should be development **not** production dependencies. Production dependencies are only those used by the Express server at runtime.

File: client/src/Components/TaskForm/TaskForm.js [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+
+		const submitHandler = (event) => {
+			event.preventDefault();
+
+			if (!formIsValid) {
+				return;
+			}
+			//need to post the task details
+			console.log("Submitted");
+			console.log(taskTitleValue);
```
Comment (textbook) [inline client/src/Components/TaskForm/TaskForm.js:42]: ...and the current handler doesn't do anything with the data.

File: client/src/Components/TaskForm/DateSelector.js [hunk 4]
```diff
@@ -0,0 +1,20 @@
+
+import React, { useState } from "react";
+import DatePicker from "react-datepicker";
+import "react-datepicker/dist/react-datepicker.css";
+
+import "./DateSelector.css";
+
+const DateSelector = () => {
+	const [startDate, setStartDate] = useState(new Date());
```
Comment (textbook) [inline client/src/Components/TaskForm/DateSelector.js:9]: How does this data get _out_ of the component? The user can edit the local state, but it doesn't seem to be connected to anything else.

File: client/src/Components/TaskForm/TaskForm.js [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+const TaskForm = ({ onSubmit }) => {
+
+    const {
+			value: taskTitleValue,
+			isValid: taskTitleIsValid,
+			hasError: taskTitleHasError,
+			valueChangeHandler: taskTitleChangeHandler,
+			inputBlurHandler: taskTitleBlurHandler,
+			reset: resetTaskTitle,
+		} = useInput(isNotEmpty);
```
Comment (textbook) [inline client/src/Components/TaskForm/TaskForm.js:26]: This is going to get very complex, with lots of aliases, as soon as you have more than one input. Maybe abstracting out your own `Input` component, that uses this hook internally, would help simplify things?

File: client/src/Components/TaskForm/DateSelector.js (hunk 4 above)
Comment (wiamnasr) [reply client/src/Components/TaskForm/DateSelector.js:9]: We will have possible pass it through a function that's passed as prop back to the component where the DateSelector is being used
is that a good option here?

File: client/src/Components/TaskForm/TaskForm.js [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+const isNotEmpty = (value) => value.trim() !== "";
+// const isEmail = (value) => value.includes("@");
+
+// Card for general styling from UI
+import Card from "../UI/Card/Card";
+import DateSelector from "./DateSelector";
+import DropdownButton from "react-bootstrap/DropdownButton";
+import DropdownItem from "react-bootstrap/esm/DropdownItem";
+
+const TaskForm = ({ onSubmit }) => {
```
[the remark this answers is not in the document]
Comment (wiamnasr) [reply client/src/Components/TaskForm/TaskForm.js:17]: Yes we will be getting this information on submission, sending it to the API as a post request with fetch, getting back the updated list of tasks and showing it back to the user/graduate

Is this approach sensible or should we consider something else, and what? we did the tasks component that will receive the taskForm component with some more tweeks and hidden form input for when the user is in edit mode, and map it according to received tasks from API

Please your feedback regarding this approach and any possible alternatives is highly useful

File: client/src/Components/TaskForm/TaskForm.js (hunk 5 above)
Comment (wiamnasr) [reply client/src/Components/TaskForm/TaskForm.js:26]: Maybe having our own hook is not so ideal for this app, are you suggesting to managing the states individually?

File: client/src/Components/TaskForm/TaskForm.js (hunk 3 above)
Comment (wiamnasr) [reply client/src/Components/TaskForm/TaskForm.js:42]: This here too will be modified (original plan that we can change if there is any suggested alternative) to include hidden form input that s visible in edit form, combining with useRef/forwardRef to  get the value, but will not update using useRef as research showed its not advised to directly manipulate the DOM with useREf

File: client/src/hooks/useInput.js [hunk 7]
```diff
@@ -0,0 +1,37 @@
+import { useState } from "react";
+
+const useInput = (validateValue) => {
```
[the remark this answers is not in the document]
Comment (wiamnasr) [reply client/src/hooks/useInput.js:3]: I learned this through a Udemy course, by this I mean creating my own custom hooks with React, we can switch to individual state management, that would be better here I think.

We did spend time as a team going through this on an example and then Pete implemented it. Even if we don't end up using it, which is rightly the case, happy that we learned this as a team

[the remark this answers is not in the document]
Comment (wiamnasr) [reply client/src/hooks/useInput.js:3]: We should definetly leave an accreditation to the Udemy Instructor! 👯

**Review by wiamnasr — COMMENTED** [review]
Pete Brilliant work, apparently we all learn super fast under pressure of deadlines!! Happy to see your efforts, if you can go ahead and remove the dependencies from the package.json and follow the other feedback points of removing the commented out code for now (maybe add the files to .gitignore ?) other than that the code is legit and glad you kept the comments as a way for other devs to follow our thinking process. If the pop-up is causing an issue, as we mentioned keep it on the side and we can move on to the mentor's side and github groups research

**Comment by wiamnasr** [comment]
Managed merging conflicts, make sure to pull changes and to run npm install again pete as I removed some duplication from package.lock-json

File: client/src/Components/TaskForm/DateSelector.js (hunk 4 above)
Comment (textbook) [reply client/src/Components/TaskForm/DateSelector.js:9]: Yes, that's a common pattern - just as this component passes the `startDate` and a callback to `setStartDate` to `DatePicker, users of this component would do the same. But until it's actually wired up it shouldn't be included - it's confusing to show something in the UI that doesn't actually _do_ anything.

File: package.json [hunk 8]
```diff
@@ -69,11 +69,15 @@
 	"dependencies": {
 		"@babel/runtime": "^7.17.0",
 		"@supabase/supabase-js": "^1.31.1",
-		"dotenv": "^16.0.0",
```
Comment (textbook) [inline package.json:72]: Something might have done wrong in syncing with main - this shouldn't remove `dotenv`, it's still used by the backend.

File: client/src/Components/TaskForm/TaskForm.js (hunk 5 above)
Comment (textbook) [reply client/src/Components/TaskForm/TaskForm.js:26]: There's no problem with hooks particularly, and managing more separate states wouldn't be helpful, but imagine what happens here when you have more than one field:
```js
const {
  value: taskTitleValue,
  isValid: taskTitleIsValid,
  hasError: taskTitleHasError,
  valueChangeHandler: taskTitleChangeHandler,
  inputBlurHandler: taskTitleBlurHandler,
  reset: resetTaskTitle,
} = useInput(isNotEmpty);
const {
  value: taskDescriptionValue,
  isValid: taskDescriptionIsValid,
  hasError: taskDescriptionHasError,
  valueChangeHandler: taskDescriptionChangeHandler,
  inputBlurHandler: taskDescriptionBlurHandler,
  reset: resetTaskDescription,
} = useInput();
const {
  value: taskStatusValue,
  isValid: taskStatusIsValid,
  hasError: taskStatusHasError,
  valueChangeHandler: taskStatusChangeHandler,
  inputBlurHandler: taskStatusBlurHandler,
  reset: resetTaskStatus,
} = useInput();
```
Every one adds _another_ six variables into the current scope, another six values you have to think about when working out what the component is doing. What if instead you pushed the state down to a sub-component, instead? For example, its interface could look like:
```js
<MyInput isValid={isNotEmpty} onChange={(value) => updateForm({ ...form, title: value })} value={form.title} />
```

File: client/src/Components/TaskForm/TaskForm.js (hunk 3 above)
Comment (textbook) [reply client/src/Components/TaskForm/TaskForm.js:42]: I'm not sure what you'd use the hidden input or ref for, my point was more that the data isn't being used - shouldn't it get e.g. POSTed somewhere?

File: client/src/Components/TaskForm/TaskForm.js (hunk 5 above)
Comment (wiamnasr) [reply client/src/Components/TaskForm/TaskForm.js:26]: makes sense, will make changes to the structure and account for the processing/creation of unnecessary variables in the scope

File: client/src/App.js (hunk 1 above)
Comment (PeterAhern) [reply client/src/App.js]: Thank you for the feedback, we had a discussion about this before the PR, but going forward we will take your suggested approach. I have removed the commented out files.

File: package.json (hunk 2 above)
Comment (PeterAhern) [reply package.json]: Thank you, I have removed these dependencies and put them in the devDependencies for now.

File: package.json (hunk 8 above)
Comment (PeterAhern) [reply package.json:72]: thank you, I have added this back in the dependencies.

**Review by wiamnasr — APPROVED** [review]

**Review by wiamnasr — APPROVED** [review]
Thank you for your efforts!

**Pull request merged**